File: iOS/BCFerries/FerryPier.swift
//
//  FerryPier.swift
//  BCFerries
//

import Foundation

/// A ferry terminal. The raw value is the enum name; `code` is the terminal code used by BC Ferries.
public enum FerryPier: String, Codable {
    case Tsawwassen
    case SwartzBay
    case HorseshoeBay
    case DepartureBay
    case DukePoint
    case Langdale
    case BowenIsland // Snug Cove
    case FulfordHarbour // Salt Spring Island

    static let ENUMS = [Tsawwassen, SwartzBay, HorseshoeBay, DepartureBay, DukePoint, Langdale, BowenIsland, FulfordHarbour]

    static let CODES: [String: FerryPier] = [
        "TSA": .Tsawwassen,
        "SWB": .SwartzBay,
        "HSB": .HorseshoeBay,
        "NAN": .DepartureBay,
        "DUK": .DukePoint,
        "LNG": .Langdale,
        "BOW": .BowenIsland,
        "FUL": .FulfordHarbour
    ]

    static let COORDS: [FerryPier: Coordinate] = [
        .Tsawwassen: Coordinate(latitude: 49.0068, longitude: -123.1290),
        .SwartzBay: Coordinate(latitude: 48.6890, longitude: -123.4105),
        .HorseshoeBay: Coordinate(latitude: 49.3757, longitude: -123.2717),
        .DepartureBay: Coordinate(latitude: 49.1935, longitude: -123.9553),
        .DukePoint: Coordinate(latitude: 49.1626, longitude: -123.8912),
        .Langdale: Coordinate(latitude: 49.4336, longitude: -123.4738),
        .BowenIsland: Coordinate(latitude: 49.3800, longitude: -123.3300),
        .FulfordHarbour: Coordinate(latitude: 48.7690, longitude: -123.4510)
    ]

    var coordinate: Coordinate {
        FerryPier.COORDS[self]!
    }

    /// The two Metro Vancouver hubs; every route has exactly one non-mainland end.
    var isMainland: Bool {
        self == .Tsawwassen || self == .HorseshoeBay
    }

    var code: String {
        FerryPier.CODES.first { $0.value == self }!.key
    }

    static func findNearest(latitude: Double, longitude: Double, piers: [FerryPier] = ENUMS) -> FerryPier? {
        let dists = piers.map { (key: $0, value: COORDS[$0]!.distance(lat: latitude, long: longitude)) }
        return dists.min { $0.value < $1.value }?.key
    }

    static func valueOf(_ s: String) -> FerryPier? {
        FerryPier(rawValue: s)
    }

    static func fromCode(_ code: String) -> FerryPier? {
        CODES[code]
    }

    func toString() -> String {
        return rawValue
    }
}

extension FerryPier: Identifiable {
    public var id: String { self.rawValue }
}

File: Android/app/src/main/java/com/lunesu/bcferries/FerryPier.kt
package com.lunesu.bcferries

/** A ferry terminal; [code] is the terminal code used by BC Ferries and the sailings feed. */
enum class FerryPier(val code: String) {
    Tsawwassen("TSA"),
    SwartzBay("SWB"),
    HorseshoeBay("HSB"),
    DepartureBay("NAN"),
    DukePoint("DUK"),
    Langdale("LNG"),
    BowenIsland("BOW"), // Snug Cove
    FulfordHarbour("FUL"); // Salt Spring Island

    val coordinate get() = COORDS.getValue(this)

    /** The two Metro Vancouver hubs; every route has exactly one non-mainland end. */
    val isMainland get() = this == Tsawwassen || this == HorseshoeBay

    companion object {
        internal val ENUMS = values()

        val COORDS = mapOf(
            Tsawwassen to Coordinate(49.0068, -123.1290),
            SwartzBay to Coordinate(48.6890, -123.4105),
            HorseshoeBay to Coordinate(49.3757, -123.2717),
            DepartureBay to Coordinate(49.1935, -123.9553),
            DukePoint to Coordinate(49.1626, -123.8912),
            Langdale to Coordinate(49.4336, -123.4738),
            BowenIsland to Coordinate(49.3800, -123.3300),
            FulfordHarbour to Coordinate(48.7690, -123.4510)
        )

        fun fromCode(code: String): FerryPier? = ENUMS.find { it.code == code }

        fun findNearest(latitude: Double, longitude: Double, piers: Array<FerryPier> = ENUMS): FerryPier? {
            return piers.minBy { it.coordinate.distance(latitude, longitude) }
        }
    }
}
